doc/CONTRIBUTING.md

```diff
@@ -70,6 +70,85 @@ or reported for next major release if BC is not possible.
 
 If you are not sure of what to do, don't hesitate to open an issue about your PR project.
 
+#### Composing the PR
+
+Nobody is perfect, and submitting perfect changes that don't need any tweaks is
+quite a rare instance. After a few iterations, your PR, which is a list of
+commits, could look like this:
+
+- replace tabs with spaces (who did this?)
+- add flux capacitor feature
+- add missing documentation (silly me)
+- add missing tests (I love TDD, but…)
+- fix typo in documentation
+- set default value as suggested
+
+Ideally, we should be able to revert any commit in your PR and not have to
+revert the others too. For instance, if we decide to revert your feature
+because it turns out to cause more problems than it solves, we should not have
+to revert the commit with the feature itself, then the commit with the
+documentation, then the commit with the tests, etc…
+
+Another thing to take into consideration is that in some hopefully rare
+instances, you will need to keep some commits separate, because they do not
+really relate to what you are doing, but the "other" thing you fixed, while
+unrelated, somehow prevents you from doing your PR, or is a minor thing you
+feel does not warrant a PR by itself, especially if this PR is conflicting with
```

Comment:
> In retrospect, I don't think it is so rare. The opposite, actually. Very very often, before diving in my feature, I have to change something phpcs-fixer cannot fix first.

Yeah, on another PRs. :stuck_out_tongue: 

> It it were a very clean project, why not but how many times do you come accross a line that makes you think "can I leave this like that, really?" when contributing? The boy-scout rule is better IMO, and we should not prevent it being applied. You're proposing several parallel PRs for that, but I think it will drive people crazy (I know it will drive me and perhaps also @ju1ius crazy).

The many automated tools we will have, cleaner the project will be.

For boy-scout rule, I'm still convince about the "If this line does not concern your change goal, does not touch it.", because it make the PR diff cleaner and we can focus only for the needed change. Some PR as enough big to not add boy-scout change on it.

And if people does not want to boy-scout the code on another PR, it's not a big mess. The code, even if not **very** clean (php-cs-fixer already did a great job), works as is. So I understand (and encourage) to cleanup code of a **modified line**, but not the rest. In anyway, the code will be touched and improved one day/

> This is why we have code review. Git masters can review what the git padawans want to merge and tell them.
> Then you need to learn about git reflog.

Maybe. But I think many contributors who already take times to help, might not be want to take more times to do commits magics. In some cases, it could be totally discouraging. Same for team member. We have already enough PR to review and merge to avoid lose time of doing this tricks yourself if the contributor does not want to.

This is also the goal of "One PR by feature".

> Finally, you say that "squash and merge" can make us avoid teaching git rebase --interactive to people, and I say they absolutely need to know about it. They are developers, and they should be able to produce clean history. And when you get them to try rebase, I'm sure they start using it on their projects, and come back better contributors. Don't fish for them with "squash and merge", try to teach them how to fish first.

Yes and no. I agree for the "educational" part. Maybe you can teach them but I think we still must use this feature. It produce a very clean history without ugly "merge commits". See: https://github.com/sonata-project/dev-kit/commits/master

So yes, teach them if you want too but don't wait 2 weeks for a user response if it's just to squash two commits.

To finalize, I think this description should not be on contributing for another reason: It's big, very big.

I think it could maybe be added later on a sub-document. The contributing have to be east to read, clear and concise so contributors know quickly what they can do and what they can not.

I'm thinking something like this on contributing doc:

* The PR goal for 99% of cases: One subject by PR + do not change not related code.
* Add a with a link to a document about "How to manage a PR commits history" if you **really** does not have the choice.

Will work on it soon.